README.md
# pwa-builder-mac README

This is the README for your extension "pwa-builder-mac". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: enable/disable this extension
* `myExtension.thing`: set to `blah` to do something

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

-----------------------------------------------------------------------------------------------------------

## Working with Markdown

**Note:** You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+CMD+V` on macOS or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux) or `Cmd+Space` (macOS) to see a list of Markdown snippets

### For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log
All notable changes to the "pwa-builder-mac" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]
- Initial release

out/extension.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require("vscode");
// Node Lib
const request = require("request");
const http = require("http");
const fs = require("fs");
const path = require("path");
const os = require("os");
// Image Generator
const AdmZip = require('adm-zip');
const FormData = require("form-data");
// Appx Packaging Lib
const Filehound = require('filehound');
// Command Line
const Q = require('q');
const exec = require('child_process').exec;
const execute = Q.nfbind(exec);
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {
    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    console.log('Congratulations, your extension "pwa-builder-mac" is now active!');
    // The command has been defined in the package.json file
    // Now provide the implementation of the command with  registerCommand
    // The commandId parameter must match the command field in package.json
    // ------------  IMAGE GENERATOR ---------------------------
    let inputBox = vscode.commands.registerCommand('extension.imageGenerator', () => {
        const apiUrl = 'http://appimagegenerator-pre.azurewebsites.net/'; // 'http://localhost:49080/';
        let extractPath = '';
        let manifestFilePath = '';
        vscode.window.showOpenDialog({
            canSelectFiles: true,
            canSelectFolders: false,
            canSelectMany: false,
            openLabel: "Select image",
            filters: {
                'Images': ['bmp', 'png', 'jpg', 'jpeg']
            }
        })
            .then(function (result) {
            let imgUrl = result[0].fsPath;
            if (imgUrl != '') {
                try {
                    const formData = new FormData();
                    formData.append('fileName', fs.createReadStream(imgUrl));
                    // Platforms input
                    vscode.window.showQuickPick(["iOS", "Windows", "Windows10", "Android", "Chrome", "Firefox"], { canPickMany: true })
                        .then(function (result) {
                        if (result != '') {
                            for (let i = 0; i < result.length; i++) {
                                formData.append('platform', (result[i]).toLowerCase());
                            }
                            inputPadding();
                        }
                        else {
                            vscode.window.showErrorMessage("No platform was chosen.");
                        }
                    });
                    // Padding input
                    let inputPadding = function () {
                        vscode.window.showInputBox({ prompt: '0 is no padding, 1 is 100% of the source image. 0.3 is a typical value for most icons', placeHolder: '0.3' })
                            .then(function (resultPadding) {
                            let predetPadding = '0.3';
                            if (resultPadding != undefined && resultPadding != null && resultPadding != '') {
                                if (parseFloat(resultPadding) < 0 || parseFloat(resultPadding) > 0) {
                                    vscode.window.showErrorMessage("Incorrect padding value. It must be between 0 and 1.");
                                }
                                else {
                                    formData.append('padding', resultPadding);
                                    inputColorOption();
                                }
                            }
                            else {
                                formData.append('padding', predetPadding);
                                inputColorOption();
                            }
                        });
                    };
                    // Image Background
                    let inputColorOption = function () {
                        vscode.window.showQuickPick(["Transparent", "Custom Color"], { canPickMany: false })
                            .then(function (result) {
                            let color = '';
                            let colorOption = '0';
                            let colorChanged = '0';
                            if (result == "Custom Color") {
                                vscode.window.showInputBox({ prompt: 'Insert the hexadecimal code. Ex: #123456' })
                                    .then(function (resultColor) {
                                    if (resultColor.substring(0, 1) == '#' && resultColor.length == 7) {
                                        color = resultColor;
                                        colorOption = '1';
                                        colorChanged = '1';
                                        formData.append('colorOption', colorOption);
                                        formData.append('colorChanged', colorChanged);
                                        formData.append('color', color);
                                        setExtractPath();
                                    }
                                    else {
                                        vscode.window.showErrorMessage("Invalid value, it must be like #123456.");
                                    }
                                });
                            }
                            else {
                                formData.append('colorOption', colorOption);
                                formData.append('colorChanged', colorChanged);
                                formData.append('color', color);
                                setExtractPath();
                            }
                        });
                    };
                    // Select of extraction path
                    let setExtractPath = function () {
                        vscode.window.showOpenDialog({
                            canSelectFiles: false,
                            canSelectFolders: true,
                            canSelectMany: false,
                            openLabel: "Select Assets Path",
                        })
                            .then(function (result) {
                            extractPath = result[0].path.substring(1);
                            let folderName = extractPath.split('/').pop();
                            if (folderName.toLowerCase() == 'assets') {
                                setManifestPath();
                            }
                            else {
                                vscode.window.showErrorMessage("Wrong folder name. It must be the Assets folder.");
                            }
                        });
                    };
                    // Select of manifest path
                    let setManifestPath = function () {
                        vscode.window.showOpenDialog({
                            canSelectFiles: true,
                            canSelectFolders: false,
                            canSelectMany: false,
                            openLabel: "Select the Manifest file",
                        })
                            .then(function (result) {
                            manifestFilePath = result[0].path.substring(1);
                            let fileNameManifest = manifestFilePath.split('/').pop(); //FileName with extension
                            let extensionFile = fileNameManifest.split('.').pop(); // Extension
                            if (extensionFile.toLowerCase() == 'json') {
                                sendToApi();
                            }
                            else {
                                vscode.window.showErrorMessage("Invalid Manifest format");
                            }
                        });
                    };
                    let sendToApi = function () {
                        console.log('send to api');
                        request.post(apiUrl + 'api/image/', {
                            body: formData,
                            headers: formData.getHeaders()
                        }, function (err, res) {
                            const resultZipUri = JSON.parse(res.body).Uri;
                            try {
                                let newFileName = 'AppImages';
                                var tmpFilePath = extractPath + "/" + newFileName + '.zip';
                                http.get(apiUrl + resultZipUri.substring(1), function (response) {
                                    response.on('data', function (data) {
                                        fs.appendFileSync(tmpFilePath, data);
                                    });
                                    response.on('end', function () {
                                        try {
                                            if (tmpFilePath) {
                                                var zip = new AdmZip(tmpFilePath);
                                                vscode.window.showInformationMessage('Extracting images');
                                                zip.extractAllTo(extractPath, true);
                                                fs.unlink(tmpFilePath);
                                                let jsonManifest = JSON.parse(fs.readFileSync(manifestFilePath, function (err, data) { if (err) {
                                                    throw err;
                                                } }));
                                                let jsonIcons = JSON.parse(fs.readFileSync(extractPath + '/icons.json', function (err, data) { if (err) {
                                                    throw err;
                                                } }));
                                                jsonManifest.icons = jsonIcons.icons;
                                                vscode.window.showInformationMessage('Updating manifest');
                                                fs.writeFileSync(manifestFilePath, JSON.stringify(jsonManifest));
                                                vscode.window.showInformationMessage("Manifest update complete.");
                                            }
                                        }
                                        catch (error) {
                                            vscode.window.showErrorMessage("error catch" + error);
                                            console.log("error catch", error);
                                        }
                                    });
                                });
                            }
                            catch (err) {
                                vscode.window.showInformationMessage("Error on Zip Request: " + err);
                            }
                        });
                    };
                }
                catch (error) {
                    vscode.window.showInformationMessage('Error: ' + error);
                }
            }
            else {
                vscode.window.showInformationMessage("No image selected");
            }
        });
    });
    // -------------------- END IMAGE GENERATOR -------------------------------------
    // ------------------------- APPX PACKAGE ------------------------------------
    let appxPackage = vscode.commands.registerCommand('extension.appxPackage', () => {
        console.log("appxPackage");
        try {
            let xmlPath;
            let filesFound;
            let manifestPath;
            let manifestJson;
            // Manifest file picker
            vscode.window.showOpenDialog({
                canSelectFiles: true,
                canSelectFolders: false,
                canSelectMany: false,
                openLabel: "Select manifest.json",
            })
                .then(function (result) {
                manifestPath = result[0].fsPath;
                manifestJson = JSON.parse(fs.readFileSync(result[0].fsPath, function (err, data) { if (err) {
                    throw err;
                } }));
                // Output folder picker where the app will be generated
                vscode.window.showOpenDialog({
                    canSelectFiles: false,
                    canSelectFolders: true,
                    canSelectMany: false,
                    openLabel: "Select destination folder",
                })
                    .then(function (result) {
                    xmlPath = result[0].fsPath;
                    // The data required by the MakeAppx is added manually
                    manifestJson.out = xmlPath;
                    manifestJson.dir = xmlPath;
                    vscode.window.showInformationMessage("Generating package. Please wait.");
                    let cmdline = "pwabuilder -m " + manifestPath + " -p mac -d " + xmlPath;
                    execute(cmdline)
                        .then(function () {
                        console.log('appname', manifestJson.short_name);
                        vscode.window.showInformationMessage("Just a little bit more...");
                        Filehound.create()
                            .glob(manifestJson.short_name)
                            .paths(xmlPath)
                            .ignoreHiddenDirectories()
                            //.ignoreHiddenFiles()
                            .find((err, htmlFiles) => {
                            if (err)
                                throw err;
                            console.log('filesfound', htmlFiles, 'error filesfound', err);
                            filesFound = htmlFiles;
                        })
                            .catch(function (err) {
                            vscode.window.showInformationMessage("Finding appxmanifest error: " + err);
                        });
                    })
                        .catch(function (cat) {
                        if (cat) {
                            throw cat;
                        }
                    });
                });
            });
        }
        catch (error) {
            vscode.window.showInformationMessage(error);
        }
    });
    // -------------------- END APPX PACKAGE ----------------------------
    // ------------------- EXECUTE APP -----------------
    let executeApp = vscode.commands.registerCommand('extension.executeApp', () => {
        vscode.window.showOpenDialog({
            canSelectFiles: true,
            canSelectFolders: false,
            canSelectMany: false,
            openLabel: "Select App",
        })
            .then(function (result) {
            try {
                let appFile = result[0].fsPath.split('/').pop();
                let path = result[0].fsPath.replace(appFile, '');
                let fileNoExt = appFile.split('.');
                Filehound.create()
                    .glob(fileNoExt[0])
                    .paths(path)
                    .find((err, htmlFiles) => {
                    if (err)
                        throw err;
                    console.log('filesfound', htmlFiles, 'error filesfound', err);
                    let comandLine = "open -F " + htmlFiles[0];
                    exec(comandLine);
                })
                    .catch(function (err) {
                    vscode.window.showInformationMessage("Finding appxmanifest error: " + err);
                });
            }
            catch (error) {
                console.log('Catch: ', error);
            }
        });
    });
    // ------------------- END EXECUTE APP -----------------
    context.subscriptions.push();
}
exports.activate = activate;
// this method is called when your extension is deactivated
function deactivate() {
}
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map

src/extension.ts
'use strict';
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';


// Node Lib
const request = require("request");
const http = require("http");
const fs = require("fs");
const path = require("path");
const os = require("os");


// Image Generator

const AdmZip = require('adm-zip');
const FormData = require("form-data");

// Appx Packaging Lib

const Filehound = require('filehound');

// Command Line
const Q = require('q');
const exec = require('child_process').exec;
const execute = Q.nfbind(exec);




// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    console.log('Congratulations, your extension "pwa-builder-mac" is now active!');

    // The command has been defined in the package.json file
    // Now provide the implementation of the command with  registerCommand
    // The commandId parameter must match the command field in package.json

    // ------------  IMAGE GENERATOR ---------------------------

    let inputBox = vscode.commands.registerCommand('extension.imageGenerator', () => {

        const apiUrl = 'http://appimagegenerator-pre.azurewebsites.net/'; // 'http://localhost:49080/';

        let extractPath = '';
        let manifestFilePath = '';

        vscode.window.showOpenDialog({
            canSelectFiles: true,
            canSelectFolders: false,
            canSelectMany: false,
            openLabel: "Select image",
            filters: {
                'Images': ['bmp', 'png', 'jpg', 'jpeg']
            }
        })
            .then(function (result: any) {

                let imgUrl = result[0].fsPath;


                if (imgUrl != '') {

                    try {

                        const formData = new FormData()

                        formData.append('fileName', fs.createReadStream(imgUrl));

                        // Platforms input

                        vscode.window.showQuickPick(["iOS", "Windows", "Windows10", "Android", "Chrome", "Firefox"], { canPickMany: true })
                            .then(function (result: any) {
                                if (result != '') {

                                    for (let i = 0; i < result.length; i++) {
                                        formData.append('platform', (result[i]).toLowerCase());
                                    }

                                    inputPadding();
                                } else {
                                    vscode.window.showErrorMessage("No platform was chosen.")
                                }
                            });



                        // Padding input
                        let inputPadding = function () {

                            vscode.window.showInputBox({ prompt: '0 is no padding, 1 is 100% of the source image. 0.3 is a typical value for most icons', placeHolder: '0.3' })
                                .then(function (resultPadding) {
                                    let predetPadding = '0.3';

                                    if (resultPadding != undefined && resultPadding != null && resultPadding != '') {
                                        if (parseFloat(resultPadding) < 0 || parseFloat(resultPadding) > 0) {
                                            vscode.window.showErrorMessage("Incorrect padding value. It must be between 0 and 1.")
                                        } else {
                                            formData.append('padding', resultPadding);
                                            inputColorOption();

                                        }
                                    } else {
                                        formData.append('padding', predetPadding);
                                        inputColorOption();

                                    }

                                })
                        };
                        // Image Background
                        let inputColorOption = function () {
                            vscode.window.showQuickPick(["Transparent", "Custom Color"], { canPickMany: false })
                                .then(function (result) {
                                    let color = '';
                                    let colorOption = '0';
                                    let colorChanged = '0';

                                    if (result == "Custom Color") {

                                        vscode.window.showInputBox({ prompt: 'Insert the hexadecimal code. Ex: #123456' })
                                            .then(function (resultColor: any) {

                                                if (resultColor.substring(0, 1) == '#' && resultColor.length == 7) {

                                                    color = resultColor;
                                                    colorOption = '1';
                                                    colorChanged = '1';

                                                    formData.append('colorOption', colorOption);
                                                    formData.append('colorChanged', colorChanged);
                                                    formData.append('color', color);

                                                    setExtractPath();
                                                } else {
                                                    vscode.window.showErrorMessage("Invalid value, it must be like #123456.")
                                                }

                                            })
                                    } else {

                                        formData.append('colorOption', colorOption);
                                        formData.append('colorChanged', colorChanged);
                                        formData.append('color', color);
                                        setExtractPath();

                                    }
                                })
                        };

                        // Select of extraction path

                        let setExtractPath = function () {
                            vscode.window.showOpenDialog({
                                canSelectFiles: false,
                                canSelectFolders: true,
                                canSelectMany: false,
                                openLabel: "Select Assets Path",

                            })
                                .then(function (result: any) {

                                    extractPath = result[0].path.substring(1);
                                    let folderName: any = extractPath.split('/').pop();
                                    if (folderName.toLowerCase() == 'assets') {
                                        setManifestPath();
                                    } else {
                                        vscode.window.showErrorMessage("Wrong folder name. It must be the Assets folder.")
                                    }


                                })
                        }
                        // Select of manifest path
                        let setManifestPath = function () {
                            vscode.window.showOpenDialog({
                                canSelectFiles: true,
                                canSelectFolders: false,
                                canSelectMany: false,
                                openLabel: "Select the Manifest file",

                            })
                                .then(function (result: any) {
                                    manifestFilePath = result[0].path.substring(1);
                                    let fileNameManifest: any = manifestFilePath.split('/').pop(); //FileName with extension
                                    let extensionFile = fileNameManifest.split('.').pop(); // Extension

                                    if (extensionFile.toLowerCase() == 'json') {
                                        sendToApi();
                                    } else {
                                        vscode.window.showErrorMessage("Invalid Manifest format");
                                    }
                                })
                        }


                        let sendToApi = function () {
                            console.log('send to api')
                            request.post(apiUrl + 'api/image/', {
                                body: formData,
                                headers: formData.getHeaders()

                            }, function (err: any, res: any) {

                                const resultZipUri = JSON.parse(res.body).Uri;

                                try {

                                    let newFileName = 'AppImages';
                                    var tmpFilePath = extractPath + "/" + newFileName + '.zip';

                                    http.get(apiUrl + resultZipUri.substring(1), function (response: any) {
                                        response.on('data', function (data: any) {

                                            fs.appendFileSync(tmpFilePath, data)

                                        });
                                        response.on('end', function () {

                                            try {
                                                if (tmpFilePath) {

                                                    var zip = new AdmZip(tmpFilePath)

                                                    vscode.window.showInformationMessage('Extracting images')

                                                    zip.extractAllTo(extractPath, true)

                                                    fs.unlink(tmpFilePath)

                                                    let jsonManifest = JSON.parse(fs.readFileSync(manifestFilePath, function (err: any, data: any) { if (err) { throw err; } }))

                                                    let jsonIcons = JSON.parse(fs.readFileSync(extractPath + '/icons.json', function (err: any, data: any) { if (err) { throw err; } }))

                                                    jsonManifest.icons = jsonIcons.icons;

                                                    vscode.window.showInformationMessage('Updating manifest')

                                                    fs.writeFileSync(manifestFilePath, JSON.stringify(jsonManifest))

                                                    vscode.window.showInformationMessage("Manifest update complete.")
                                                }

                                            } catch (error) {
                                                vscode.window.showErrorMessage("error catch" + error)
                                                console.log("error catch", error)
                                            }
                                        })
                                    });

                                }

                                catch (err) {
                                    vscode.window.showInformationMessage("Error on Zip Request: " + err)
                                }
                            });
                        };

                    } catch (error) {
                        vscode.window.showInformationMessage('Error: ' + error)
                    }
                } else {
                    vscode.window.showInformationMessage("No image selected")
                }

            })


    });
    // -------------------- END IMAGE GENERATOR -------------------------------------

    // ------------------------- APPX PACKAGE ------------------------------------

let appxPackage = vscode.commands.registerCommand('extension.appxPackage', () => {
    console.log("appxPackage")

    try {
        let xmlPath:any;
        let filesFound:any;
        let manifestPath:any;
        let manifestJson:any;

        // Manifest file picker

        vscode.window.showOpenDialog({
            canSelectFiles: true,
            canSelectFolders: false,
            canSelectMany: false,
            openLabel: "Select manifest.json",

        })
        .then(function(result:any){
        manifestPath = result[0].fsPath;
        manifestJson = JSON.parse(fs.readFileSync(result[0].fsPath, function(err:any, data:any){if(err){throw err;}}));

            // Output folder picker where the app will be generated
            vscode.window.showOpenDialog({
                canSelectFiles: false,
                canSelectFolders: true,
                canSelectMany: false,
                openLabel: "Select destination folder",

            })
            .then(function(result:any){
                xmlPath = result[0].fsPath;
                // The data required by the MakeAppx is added manually
                manifestJson.out = xmlPath;
                manifestJson.dir = xmlPath;

                vscode.window.showInformationMessage("Generating package. Please wait.")
                let cmdline = "pwabuilder -m " + manifestPath + " -p mac -d " + xmlPath
                execute(cmdline)
                .then(
                    function () {
                        console.log('appname', manifestJson.short_name)
                        vscode.window.showInformationMessage("Just a little bit more...")

                        Filehound.create()
                        .glob(manifestJson.short_name)
                        .paths(xmlPath)
                        .ignoreHiddenDirectories()
                        //.ignoreHiddenFiles()
                            .find((err:any, htmlFiles:any) => {
                                if (err) throw err;
                                console.log('filesfound',htmlFiles, 'error filesfound',err)
                                filesFound = htmlFiles;
                            })
                            .catch(function (err:any) {
                                vscode.window.showInformationMessage("Finding appxmanifest error: " + err)
                            });

                    }
                )
                .catch(function (cat:any) {
                    if (cat) {
                        throw cat
                    }
                });

            })
        })
    } catch (error) {
        vscode.window.showInformationMessage(error)
    }


});

// -------------------- END APPX PACKAGE ----------------------------

// ------------------- EXECUTE APP -----------------
let executeApp = vscode.commands.registerCommand('extension.executeApp', () => {

    vscode.window.showOpenDialog({
        canSelectFiles: true,
        canSelectFolders: false,
        canSelectMany: false,
        openLabel: "Select App",

    })
    .then(function(result:any){

        try{

            let appFile = result[0].fsPath.split('/').pop()
            let path = result[0].fsPath.replace(appFile,'')

            let fileNoExt = appFile.split('.');

            Filehound.create()
            .glob(fileNoExt[0])
            .paths(path)

                .find((err:any, htmlFiles:any) => {
                    if (err) throw err;
                    console.log('filesfound',htmlFiles, 'error filesfound',err)

                    let comandLine = "open -F " + htmlFiles[0]

                    exec(comandLine)


                })
                .catch(function (err:any) {
                    vscode.window.showInformationMessage("Finding appxmanifest error: " + err)
                });





        } catch (error) {
            console.log('Catch: ', error)
        }
    })
});

// ------------------- END EXECUTE APP -----------------

    context.subscriptions.push();
}

// this method is called when your extension is deactivated
export function deactivate() {
}